File: lib/domain/usecases/task/complete_task_usecase.dart
import '../../entities/commit.dart' as entity;
import '../../entities/enums.dart';
import '../../entities/task.dart';
import '../../repositories/i_commit_repository.dart';
import '../../repositories/i_task_repository.dart';

/// 完成任务用例
class CompleteTaskUseCase {
  CompleteTaskUseCase(
    this._taskRepository,
    this._commitRepository,
  );

  final ITaskRepository _taskRepository;
  final ICommitRepository _commitRepository;

  Future<Task> execute(String taskId) async {
    final task = await _taskRepository.getById(taskId);
    if (task == null) {
      throw ArgumentError('任务不存在: $taskId');
    }

    if (task.status == TaskStatus.done) {
      throw StateError('任务已完成');
    }

    final updated = task.copyWith(
      status: TaskStatus.done,
      completedAt: DateTime.now(),
    );

    final savedTask = await _taskRepository.update(updated);

    // 记录完成提交
    await _commitRepository.create(
      entity.Commit.create(
        taskId: savedTask.id,
        branchId: savedTask.branchId,
        message: '完成任务: ${savedTask.title}',
        type: CommitType.complete,
      ),
    );

    return savedTask;
  }
}

File: web/src/application/usecases/task/complete-task-usecase.ts
import { Commit } from '../../../domain/entities/commit';
import { CommitType, TaskStatus } from '../../../domain/entities/enums';
import { Task } from '../../../domain/entities/task';
import { ICommitRepository } from '../../../domain/repositories/i-commit-repository';
import { ITaskRepository } from '../../../domain/repositories/i-task-repository';
import { validateNotEmpty } from '../../../core/utils/validators';

export class CompleteTaskUseCase {
  constructor(
    private taskRepo: ITaskRepository,
    private commitRepo: ICommitRepository
  ) {}

  async execute(id: string): Promise<Task> {
    validateNotEmpty(id, '任务 ID');
    const task = await this.taskRepo.getById(id);
    if (!task) {
      throw new Error('任务不存在');
    }
    if (task.isCompleted) {
      throw new Error('任务已经完成');
    }
    const completed = task.copyWith({
      status: TaskStatus.done,
      completedAt: new Date(),
    });
    const result = await this.taskRepo.update(completed);
    const commit = Commit.create(
      result.id,
      result.branchId,
      `完成任务 ${result.title}`,
      CommitType.complete
    );
    await this.commitRepo.create(commit);
    return result;
  }
}
